README.md
# KafkaLoadGenerator

This utility is used to generate load on Kafka. In the simplest form it can generate x messages per second where x is configurable. This utility has currently been enhanced to generate a near real scenario which follows the life cycle of an event and its child entities ( markets and selections).

The life cycle is as follows.

* Event/Market/Selection Creation
* Event/Market/Selection Updates
* Events kick off (isEventStarted = true)
* Event/Market/Selection Updates
* Selections/Markets/Events Resulted and Settled

    
## Building the Project

Use Maven build to build the jar. Once built the following runnable jar will be created

    kafka-load-generator-0.0.1-SNAPSHOT-jar-with-dependencies.jar
    

### Setting up the runtime

For setting up a runtime create a folder and copy the jar in it. As an example let us assume the root folder is /opt/kafkaloadgenerator.
The following folder structure will be needed
```
/opt/kafkaloadgenerator
/opt/kafkaloadgenerator/config
/opt/kafkaloadgenerator/data
```

/opt/kafkaloadgenerator will contain the jar.

/opt/kafkaloadgenerator/config contains the main configuration file and the Logger config file

```
kafkaclient.properties
Logger.properties
```

/opt/kafkaloadgenerator/data contains all the template json files ( for event/market/selection creates, event inplays and other updates)

```
PushTest_EventCreate.json
PushTest_EventInplay.json
PushTest_EventResult.json
PushTest_EventUpdate.json
PushTest_MarketCreate.json
PushTest_MarketResult.json
PushTest_MarketUpdate.json
PushTest_SelectionCreate.json
PushTest_SelectionResult.json
PushTest_SelectionUpdate.json
PushTest_SelectionUpdatePrice.json
```
All the template json files have certain key fields that will be replaced by the test utility at runtime

```
${event} will be replaced by eventKey
${market} will be replaced by the marketKey
${selection} will be replaced by the selectionKey
${toplevel} will be replaced by catergory class and type details
${messageid} will be replaced by a unique number per request
${recordModifiedTime} will be replaced by the time the message is generated and sent to Kafka
-9999 will be replaced by the displayOrder
-1111 will be replaced by a random price numerator
-2222 will be replaced by a random price denominator
3333.33 will be replaced by a random decimal price
```

## kafkaclient.properties configuration

Details of each parameter as given below

```
[client.TPS] - This is the message rate per second that you want to generate.
[client.maxtime] - This is duration of the test in seconds.

[client.producer.payload.dir] - This is the folder where the template json data is.
[client.producer.payload.fileprefix] - All templace files must have this prefix. PushTest_ as an example

[client.producer.payload.toplevels.count] - This is the total number of sport ot parent combination that will be used. In the below examples 6 top level cominations were used.
[client.producer.payload.toplevels.0] - The value has a format categoryid-classid-typeid:Percentage OF events for this sport as shown below in following examples
[client.producer.payload.toplevels.1]=34-266-10449:20   - Here Tennis ( 34 ) , All Tennis (266) and Wimbledon ( 10449) are the hierarchy and 20% of the events will be traded for this combination.
[client.producer.payload.toplevels.2]=6-13-35:40
[client.producer.payload.toplevels.3]=10-58-162:40
[client.producer.payload.toplevels.4]=31-259-2703:40
[client.producer.payload.toplevels.5]=18-195-1260:40


[client.producer.payload.eventid.range] - This is the eventid range. Keey the range to 250 and use 3 digit numbers. (e.g. 200-450) 
[client.producer.payload.marketid.prefix.range] - MarketId prefix range . the actual market id will be this range number + the event id. (e.g 70-72 will generate 3 markets under event event.  with the IDs 70200,71200 and 72200 for the event id 200) 
[client.producer.payload.selectionid.prefix.range] - Selection ID Prefix range. For the POV keep this to 3 (e.g 80-82 wil generate 8070200,8170200,8270200 for market 70200 and for event 200)

[client.producer.inplaytime] - This is the time in seconds after which events are made inplay (Recommended value 5 minutes)
[client.producer.resulttime] - This is the time in seconds after which results will be sent. This should be a little before the total time of test (Recommended value is 25 minutes for a 30 minute test)

#Kafka
[kafka.bootstrap.servers] - This is the Kafka Bootstrap Server URL
[kafka.topic] - This is the delta topic on which our Kafka Streams will be listening on.

[kafka.listener.groupName] This utility will also have a consumer which shows you the latency of consumption. This property is the consumer group name. Recommended value netty.loadtest.consumer


Please leave all other properties as is

```

### Running the service

From /opt/kafkaloadgenerator execute the following command

```
java -Dconfig=./config -jar kafka-load-generator-0.0.1-SNAPSHOT-jar-with-dependencies.jar ./config/kafkaclient.properties
```

If the utility is running fine you will see the following in the logs 

```
Stats: TPS(avg):1044, TPS(last):1012, TIME: 1341, sent:1400, ok:1400, error:0(c/t/d/o: 0/0/1598/0), avg_latency:7
Stats: TPS(avg):1024, TPS(last):1000, TIME: 2344, sent:2400, ok:2400, error:0(c/t/d/o: 0/0/1598/0), avg_latency:6
Stats: TPS(avg):1016, TPS(last):1000, TIME: 3348, sent:3400, ok:3400, error:0(c/t/d/o: 0/0/1598/0), avg_latency:5
Stats: TPS(avg):1011, TPS(last):1000, TIME: 4352, sent:4400, ok:4400, error:0(c/t/d/o: 0/0/1598/0), avg_latency:5
Stats: TPS(avg):1008, TPS(last):1000, TIME: 5356, sent:5400, ok:5400, error:0(c/t/d/o: 0/0/1598/0), avg_latency:5
Stats: TPS(avg):1006, TPS(last):1000, TIME: 6359, sent:6400, ok:6400, error:0(c/t/d/o: 0/0/1598/0), avg_latency:5

which shows the tps and the latency of the internal consumer. 
```

CHANGELOG.md
# Change Log

## [0.1.2](https://github.com/kaazing/KafkaLoadGenerator/tree/0.1.2) (2018-03-07)
[Full Changelog](https://github.com/kaazing/KafkaLoadGenerator/compare/0.1.1...0.1.2)

**Merged pull requests:**

- Prevent den price from going above PRICE\_DEN\_LIMIT [\#3](https://github.com/kaazing/KafkaLoadGenerator/pull/3) ([robinzimmermann](https://github.com/robinzimmermann))

## [0.1.1](https://github.com/kaazing/KafkaLoadGenerator/tree/0.1.1) (2018-03-07)
[Full Changelog](https://github.com/kaazing/KafkaLoadGenerator/compare/0.1.0...0.1.1)

**Merged pull requests:**

- fix divide by zero [\#2](https://github.com/kaazing/KafkaLoadGenerator/pull/2) ([sidda-at-kaazing](https://github.com/sidda-at-kaazing))

## [0.1.0](https://github.com/kaazing/KafkaLoadGenerator/tree/0.1.0) (2018-03-01)
**Merged pull requests:**

- Kafka load generator [\#1](https://github.com/kaazing/KafkaLoadGenerator/pull/1) ([sidda-at-kaazing](https://github.com/sidda-at-kaazing))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

src/main/java/com/lc/df/kafka/client/Utils.java
package com.lc.df.kafka.client;

import java.io.File;
import java.io.IOException;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.io.filefilter.FileFilterUtils;

public class Utils
{
	private static final Random RANDOM = new Random(System.currentTimeMillis());
	private static final int PRICE_NUM_LIMIT = 500;
	private static final int PRICE_DEN_LIMIT = 1000;
	public static final List<PayloadConfig> PAYLOAD_LIST = new ArrayList<PayloadConfig>();
	
	private static String EVT_CREATE_PAYLOAD = null;
	private static String MKT_CREATE_PAYLOAD = null;
	private static String SEL_CREATE_PAYLOAD = null;
	private static String EVT_INPLAY_PAYLOAD = null;
	private static String EVT_RESULT_PAYLOAD = null;
	private static String MKT_RESULT_PAYLOAD = null;
	private static String SEL_RESULT_PAYLOAD = null;
	
	public volatile static int payloadListSize = 0;
	private volatile static boolean createsCompleted = false; 
	private volatile static long currentEventID = KafkaClientConfig.eventid_min;
	private volatile static int currentMarketPrefixID = -1;
	private volatile static int currentSelectionPrefixID = -1;
	
	private volatile static boolean sendInplay = false;
	private volatile static boolean inplayComplete = false;
	private volatile static boolean sendResults = false;
	private volatile static boolean resultsComplete = false;
	
	private static final Map<String, String> eventParentsMap = new HashMap<String,String>();
	
	public static String readFile(String file) {
		return readFile(new File(file));
	}

	public static String readFile(File file) {
		try {
			return FileUtils.readFileToString(file);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			Logger.logInfoMessage("ERROR reading life");
		}
		return null;
	}
	public static List<String> readLines(File f)
	{
		try {
			return FileUtils.readLines(f);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return new ArrayList<String>();
	}
	public static String getCorrID(String value)
	{
    	try 
    	{
    		String msgIDString = "messageID";
			int indexOfMsgID = value.indexOf(msgIDString);
			if (indexOfMsgID > -1)
			{
				int indexOfComma = value.indexOf(",", indexOfMsgID);
				String str = value.substring(indexOfMsgID + msgIDString.length() + 2, indexOfComma);
				return str.replaceAll("\"", "").replaceAll(":", "").trim();
			}
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    	return "";
	}
	public static String getNewPayload(String payload,String eventid,String marketid,String selectionid,String marketIDPrefix,String selectionIdPrefix,String messageid)
	{
		try 
		{
			String currentTimeStamp = Long.toString(System.currentTimeMillis());
			String pl = payload;
			String toplevel = eventParentsMap.get(eventid);
			if (toplevel == null){
				toplevel = "";
			}
			if (selectionid != null)
			{
				int num = getRandomIntInRange(PRICE_NUM_LIMIT);
				int den = getRandomIntInRange(PRICE_DEN_LIMIT - 1) + 1; // prevent denominator of zero
				double decimal = 1 + ((double)num/(double)den);
				DecimalFormat f = new DecimalFormat("##.00");
				pl = payload.replaceAll("\\$\\{selection\\}", selectionid)
						.replaceAll("\\$\\{market\\}", marketid)
						.replaceAll("-9999",selectionIdPrefix)
						.replaceAll("-1111", ""+num)
						.replaceAll("-2222", ""+den)
						.replaceAll("3333.33", ""+f.format(decimal));
			}
			else if (marketid != null)
			{
				pl = payload.replaceAll("\\$\\{market\\}", marketid)
							.replaceAll("-9999",marketIDPrefix);
			}
			return 	pl.replaceAll("\\$\\{toplevel\\}", toplevel)
					.replaceAll("\\$\\{event\\}", eventid)
					.replaceAll("-9999", eventid)
					.replaceAll("\\$\\{recordModifiedTime\\}", currentTimeStamp)
					.replaceAll("\\$\\{messageid\\}",messageid);

		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return "";
	}
	
	public static int getRandomIntInRange(int range)
	{
		return RANDOM.nextInt(range);
	}
	public static void loadPayloadAndIDs()
	{
		if (KafkaClientConfig.payload_file != null && KafkaClientConfig.payload_file.length() > 0) 
		{
			String fileContent = Utils.readFile(KafkaClientConfig.payload_file);
			if (fileContent != null && fileContent.length() > 0)
			{
				PAYLOAD_LIST.add(new PayloadConfig("General",fileContent));
			}
		} else 
		{
			File dir = new File(KafkaClientConfig.payload_dir);
			
			Collection<File> payloadFiles = FileUtils.listFiles(dir,FileFilterUtils.prefixFileFilter(KafkaClientConfig.payload_fileprefix),null);
			for (File file : payloadFiles) 
			{
				String payload = Utils.readFile(file);
				String key = FilenameUtils.getBaseName(file.getName());
				if (key.contains("EventCreate"))
				{
					EVT_CREATE_PAYLOAD = payload;
				}
				if (key.contains("EventUpdate"))
				{
					PAYLOAD_LIST.add(new PayloadConfig("E",payload));
				}
				if (key.contains("EventInplay"))
				{
					EVT_INPLAY_PAYLOAD = payload;
				}
				if (key.contains("EventResult"))
				{
					EVT_RESULT_PAYLOAD = payload;
				}
				if (key.contains("MarketCreate"))
				{
					MKT_CREATE_PAYLOAD = payload;
				}
				if (key.contains("MarketUpdate"))
				{
					PAYLOAD_LIST.add(new PayloadConfig("M", payload));
				}
				if (key.contains("MarketResult"))
				{
					MKT_RESULT_PAYLOAD = payload;
				}
				if (key.contains("SelectionCreate"))
				{
					SEL_CREATE_PAYLOAD = payload;
				}
				if (key.contains("SelectionUpdate"))
				{
					if (key.contains("Price"))
					{
						for (int i = 0; i < KafkaClientConfig.selection_priceupdate_multiplier; i++) {
							PAYLOAD_LIST.add(new PayloadConfig("S",payload));
						}
					}
					else
					{
						PAYLOAD_LIST.add(new PayloadConfig("S",payload));
					}
				}
				if (key.contains("SelectionResult"))
				{
					SEL_RESULT_PAYLOAD = payload;
				}
				
			}
		}

		for (long i = KafkaClientConfig.eventid_min; i <= KafkaClientConfig.eventid_max; i++) {
			eventParentsMap.put(Long.toString(i),KafkaClientConfig.topLevelList.removeFirst());
		}
		synchronized (PAYLOAD_LIST) 
		{
			payloadListSize = PAYLOAD_LIST.size();
		}
	}

	private synchronized static KafkaPayload getCreatePayload(String correlationId)
	{
		if (createsCompleted || currentEventID > KafkaClientConfig.eventid_max)
		{
			Logger.logInfoMessage("Creates Completed.....");
			return null;
		}
		if (currentMarketPrefixID == -1 && currentSelectionPrefixID == -1)
		{
			String eventID = Long.toString(currentEventID);
			currentMarketPrefixID = KafkaClientConfig.marketid_prefix_min;
			return new KafkaPayload(eventID,"event-create",getNewPayload(EVT_CREATE_PAYLOAD, eventID, null, null,null,null, correlationId));
		}
		else if (currentMarketPrefixID > -1 && currentSelectionPrefixID == -1)
		{
			String eventID = Long.toString(currentEventID);
			String marketIDPrefixString = Integer.toString(currentMarketPrefixID);
			String marketID = Integer.toString(currentMarketPrefixID) + eventID;
			currentSelectionPrefixID = KafkaClientConfig.selectionid_prefix_min;
			return new KafkaPayload(eventID,"market-create",getNewPayload(MKT_CREATE_PAYLOAD, eventID, marketID, null,marketIDPrefixString,null, correlationId));
		}
		else if (currentMarketPrefixID > -1 && currentSelectionPrefixID > -1)
		{
			String eventID = Long.toString(currentEventID);
			String marketIDPrefixString = Integer.toString(currentMarketPrefixID);
			String marketID = marketIDPrefixString + eventID;
			String selectionIDPrefixString = Integer.toString(currentSelectionPrefixID);
			String selectionID = selectionIDPrefixString + marketID;
			currentSelectionPrefixID++;
			if (currentSelectionPrefixID > KafkaClientConfig.selectionid_prefix_max)
			{
				currentSelectionPrefixID = -1;
				currentMarketPrefixID++;
				if (currentMarketPrefixID > KafkaClientConfig.marketid_prefix_max)
				{
					currentMarketPrefixID = -1;
					currentEventID++;
				}
			}
			createsCompleted = currentEventID > KafkaClientConfig.eventid_max;
			return new KafkaPayload(eventID,"selection-create",getNewPayload(SEL_CREATE_PAYLOAD, eventID, marketID, selectionID,marketIDPrefixString,selectionIDPrefixString, correlationId));
		}
		return null;
	}
	private synchronized static KafkaPayload getInplayPayload(String correlationId)
	{
		if (inplayComplete || currentEventID > KafkaClientConfig.eventid_max)
		{
			Logger.logInfoMessage("Inplay Completed.....");
			return null;
		}
		String eventID = Long.toString(currentEventID++);
		inplayComplete = currentEventID > KafkaClientConfig.eventid_max;
		return new KafkaPayload(eventID,"event-inplay",getNewPayload(EVT_INPLAY_PAYLOAD, eventID, null, null,null,null, correlationId));
	}
	private synchronized static KafkaPayload getResultsPayload(String correlationId)
	{
		if (resultsComplete || currentEventID > KafkaClientConfig.eventid_max)
		{
			Logger.logInfoMessage("Results Completed.....");
			return null;
		}
		if (currentMarketPrefixID == -1 && currentSelectionPrefixID == -1)
		{
			String eventID = Long.toString(currentEventID++);
			if (currentEventID > KafkaClientConfig.eventid_max){
				resultsComplete = true;				
			}
			else {
				currentMarketPrefixID = KafkaClientConfig.marketid_prefix_min;
				currentSelectionPrefixID = KafkaClientConfig.selectionid_prefix_min;
			}
			return new KafkaPayload(eventID,"event-result",getNewPayload(EVT_RESULT_PAYLOAD, eventID, null, null,null,null, correlationId));
		}
		else if (currentMarketPrefixID > -1 && currentSelectionPrefixID == -1)
		{
			String eventID = Long.toString(currentEventID);
			String marketID = Integer.toString(currentMarketPrefixID++) + eventID;
			if (currentMarketPrefixID > KafkaClientConfig.marketid_prefix_max){
				currentMarketPrefixID = -1;
			}
			else{
				currentSelectionPrefixID = KafkaClientConfig.selectionid_prefix_min;
			}
			return new KafkaPayload(eventID,"market-result",getNewPayload(MKT_RESULT_PAYLOAD, eventID, marketID, null,null,null, correlationId));
		}		
		else if (currentMarketPrefixID > -1 && currentSelectionPrefixID > -1)
		{
			String eventID = Long.toString(currentEventID);
			String marketID = Integer.toString(currentMarketPrefixID) + eventID;
			String selectionID = Integer.toString(currentSelectionPrefixID) + marketID;
			currentSelectionPrefixID++;
			if (currentSelectionPrefixID > KafkaClientConfig.selectionid_prefix_max){
				currentSelectionPrefixID = -1;
			}
			return new KafkaPayload(eventID,"selection-result",getNewPayload(SEL_RESULT_PAYLOAD, eventID, marketID, selectionID,null,null, correlationId));
		}

		return null;
	}
    private static long getRandomLongId(long min, long max)
    {
        if (min < max)
        {
            return ThreadLocalRandom.current().nextLong(min, max);
        }
        else
        {
            // Handle the case where the min and the max are the same.
            return min;
        }
    }
    private static int getRandomIntId(int min, int max)
    {
        if (min < max)
        {
            return ThreadLocalRandom.current().nextInt(min, max);
        }
        else
        {
            // Handle the case where the min and the max are the same.
            return min;
        }
    }
	public static KafkaPayload getRandomPayload(String correlationId)
	{
		if (resultsComplete){
			return new KafkaPayload("STOP","STOP", "STOP");
		}
		else if (!createsCompleted) 
		{
			return getCreatePayload(correlationId);
		}
		else if (sendInplay && !inplayComplete){
			return getInplayPayload(correlationId);
		}
		else if (sendResults && !resultsComplete){
			return getResultsPayload(correlationId);
		}
		else{
			int plID = getRandomIntInRange(payloadListSize);
			PayloadConfig kPayload = PAYLOAD_LIST.get(plID >0 ? plID - 1:0);
			if (kPayload != null)
			{
				String payload = kPayload.getPayload();
				if ("E".equals(kPayload.getType()))
				{
				    long eID = getRandomLongId(KafkaClientConfig.eventid_min, KafkaClientConfig.eventid_max);
					String eventKey = Long.toString(eID);
					String newPayload = getNewPayload(payload, eventKey, null,null,null,null, correlationId);
					return new KafkaPayload(eventKey,"event-update",newPayload);
				}
				if ("M".equals(kPayload.getType()))
				{
				    long eID = getRandomLongId(KafkaClientConfig.eventid_min, KafkaClientConfig.eventid_max);
				    int mID = getRandomIntId(KafkaClientConfig.marketid_prefix_min, KafkaClientConfig.marketid_prefix_max);
					String eventKey = Long.toString(eID);
					String marketKey = Integer.toString(mID) + eventKey;
					String newPayload = getNewPayload(payload, eventKey, marketKey,null,null,null, correlationId);
					return new KafkaPayload(eventKey,"market-update",newPayload);
				}
				if ("S".equals(kPayload.getType()))
				{
				    long eID = getRandomLongId(KafkaClientConfig.eventid_min, KafkaClientConfig.eventid_max);
				    int mID = getRandomIntId(KafkaClientConfig.marketid_prefix_min, KafkaClientConfig.marketid_prefix_max);
				    int sID = getRandomIntId(KafkaClientConfig.selectionid_prefix_min, KafkaClientConfig.selectionid_prefix_max);
					String eventKey = Long.toString(eID);
					String marketKey = Integer.toString(mID) + eventKey;
					String selectionKey = Integer.toString(sID) + marketKey;
					String newPayload = getNewPayload(payload, eventKey, marketKey,selectionKey,null,null, correlationId);
					return new KafkaPayload(eventKey,"selection-update",newPayload);
				}
			}			
		}
		return null;
	}
	public static synchronized void startInplay(){
		if (!sendInplay){
			sendInplay = true;
			inplayComplete = false;
			currentEventID = KafkaClientConfig.eventid_min;
		}
	}
	public static synchronized void startResults(){
		if (!sendResults){
			sendResults = true;
			resultsComplete = false;
			currentEventID = KafkaClientConfig.eventid_min;
			currentMarketPrefixID = KafkaClientConfig.marketid_prefix_min;
			currentSelectionPrefixID = KafkaClientConfig.selectionid_prefix_min;
		}
	}
}
